src/main/java/edu/kit/compiler/optimizations/Optimizer.java

```diff
@@ -0,0 +1,106 @@
+package edu.kit.compiler.optimizations;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+import firm.Dump;
+import firm.Graph;
+import firm.Program;
+
+public final class Optimizer {
+
+    private final List<Optimization.Global> globalOptimizations;
+    private final List<Optimization.Local> localOptimizations;
+    private final boolean dumpGraphs;
+
+    public Optimizer(List<Optimization.Global> globalOptimizations,
+            List<Optimization.Local> localOptimizations, boolean dumpGraphs) {
+        this.globalOptimizations = List.copyOf(globalOptimizations);
+        this.localOptimizations = List.copyOf(localOptimizations);
+        this.dumpGraphs = dumpGraphs;
+    }
+
+    /**
+     * Run all global and local optimizations in turns until a fix point is
+     * reached.
+     */
+    public void optimize() {
+        dumpGraphsIfEnabled("raw");
+
+        var changeSet = getAllGraphs();
+        boolean hasChanged;
+
+        do {
+            var callGraph = CallGraph.create();
+
+            // ? maybe only run global opts once per iteration
+            Set<Graph> newChanges;
+            do {
+                newChanges = optimizeGlobal(callGraph);
+                changeSet.addAll(newChanges);
+            } while (!newChanges.isEmpty());
+
+            hasChanged = optimizeLocal(changeSet);
```

Comment:
Is there a reason why we run the global optimizations first? I imagine local optimizations are more efficient, given that they have less information to evaluate (only the local graph). So we could first try to optimize as much as possible locally and then apply global optimizations afterwards (in rounds, of course).

Reply:
I've swapped the order of local and global optimizations, there was no deeper reasoning behind the old order. You're probably right that optimizing locally as far as possible first, may be more efficient.
Regarding your point about local optimizations having less information, that may be subject to change (see discussion thread).